web3/middleware/filter.py

```diff
@@ -0,0 +1,131 @@
+
+    def __next__(self):
+        for _start in self._count_from:
+            if self.stop and _start > self.stop:
+                raise StopIteration
+            _end = _start + (self.step - 1)
+            if self.stop and self.stop < _end:
+                _end = self.stop
+            return _start, _end
+
+
+def unload_latest_blocks(web3, fromBlock, toBlock):
+    _fromBlock = fromBlock
+
+    current_block = int(web3.eth.blockNumber)
+    if toBlock is not None and toBlock < current_block:
+        return fromBlock, toBlock
+
+    while toBlock is None or toBlock > current_block:
+        # return blocks up to current block
+        yield _fromBlock, current_block
+        _fromBlock = current_block + 1
+        current_block = int(web3.eth.blockNumber)
+        # no new blocks
+        if current_block < _fromBlock:
+            yield (None, None)
+        # reach stop_block after being passed by current_block
+        if toBlock is not None and current_block > toBlock and _fromBlock < toBlock:
+            return _fromBlock, toBlock
+
+
+def request_getLogs(
+        web3,
+        startBlock,
+        stopBlock,
+        address,
+        topics,
+        chunk_size):
+    block_chunker = RangeCounter(startBlock, stopBlock)
+    for chunk in block_chunker:
+        params = {
+            'fromBlock': chunk[0],
+            'toBlock': chunk[1],
+            'address': address,
+            'topics': topics
+        }
+        yield web3.eth.getLogs(params)
+
+
+def log_filter(
+        web3,
+        fromBlock=None,
+        toBlock=None,
+        address=None,
+        topics=None,):
+
+    if not fromBlock or fromBlock == 'latest':
+        _fromBlock = int(web3.eth.blockNumber)
+    else:
+        _fromBlock = fromBlock
+
+    if _fromBlock > int(web3.eth.blockNumber):
+        raise ValueError("fromBlock ({0}) has not yet been mined".format(fromBlock))
+
+    if toBlock == 'latest':
+        _toBlock = None
+    else:
+        _toBlock = toBlock
+
+    for start, stop in unload_latest_blocks(web3, _fromBlock, _toBlock):
+        if None in (start, stop):
+            yield []
+            #  Temporary infinite loop blocker
+            break
+
+        yield concat(request_getLogs(
+            web3,
+            start,
+            stop,
+            address,
+            topics,
+            chunk_size=5))
+
+
+def filter_middleware(make_request, web3):
+    filters = {}
+    filter_id_counter = map(to_hex, itertools.count())
+
+    def middleware(method, params):
+        if method in (
+            #  'eth_newPendingBlockFilter',
+            #  'eth_newBlockFilter',
+                'eth_newFilter',):
+
+            filter_id = next(filter_id_counter)
+
+            if method == 'eth_newFilter':
+                _filter = log_filter(web3, **params[0])
+
+            filters[filter_id] = _filter
+            return {'result': filter_id}
+        elif method == 'eth_getFilterChanges' or method == 'eth_getFilterLogs':
```

Comment:
It looks like `eth_getFilterLogs` is supposed to have it's own separate implementation? (based on your PR checklist and the differences in their respective functionality)